nemo_aligner/models/nlp/gpt/megatron_gpt_spin_model.py

```diff
@@ -0,0 +1,524 @@
+        if self.distributed_adam_offload_manager is None:
+
+            self.distributed_adam_offload_manager = (
+                offload_distributed_adam(self._optimizer.state_dict(state_dict_format=1, gather_on_root=False))
+                if self.to_offload_adam_states and self.with_distributed_adam
+                else nullcontext()
+            )
+
+            # offload onto cpu
+            self.distributed_adam_offload_manager.__enter__()
+
+    def onload_adam_states(self):
+        if self.distributed_adam_offload_manager is not None:
+            # load back onto GPU
+            self.distributed_adam_offload_manager.__exit__(None, None, None)
+
+        self.distributed_adam_offload_manager = None
+
+    def sharded_state_dict(self, prefix: str = ""):
+        sharded_state_dict = super().sharded_state_dict(prefix=prefix)
+
+        # add in the reference policy weights
+        if self.ref_policy_state_dict is not None:
+            sharded_state_dict["reference_policy"] = make_sharded_tensors_for_checkpoint(
+                self.ref_policy_state_dict, prefix
+            )
+
+        return sharded_state_dict
+
+    def on_load_checkpoint(self, checkpoint) -> None:
+        """LightningModule hook:
+        https://pytorch-lightning.readthedocs.io/en/stable/common/lightning_module.html#on-load-checkpoint
+        """
+
+        # mcore uses distributed checkpointing
+        if self.mcore_gpt:
+            # checkpoint keys: ['epoch', 'global_step', 'pytorch-lightning_version', 'state_dict', 'loops', 'callbacks', 'optimizer_states', 'lr_schedulers', 'hparams_name', 'hyper_parameters']
+            if "state_dict" in checkpoint and checkpoint["state_dict"]:
+                for index, module in enumerate(self.get_gpt_module_list()):
+                    if parallel_state.get_virtual_pipeline_model_parallel_world_size() is not None:
+                        checkpoint_state_dict = checkpoint["state_dict"][f"model_{index}"]
+                    else:
+                        checkpoint_state_dict = checkpoint["state_dict"]
+                    # checkpoint_state_dict has "model." but module does not so we need to remove it when loading
+                    checkpoint_state_dict = {
+                        key.replace("model.", ""): checkpoint_state_dict.pop(key)
+                        for key in list(checkpoint_state_dict.keys())
+                    }
+                    ref_policy = checkpoint_state_dict.pop("reference_policy", None)
+                    if ref_policy is not None:
+                        self.ref_policy_state_dict = ref_policy
+                    module.load_state_dict(checkpoint_state_dict, strict=True)
+            else:
+                # when restoring a distributed checkpoint from a ptl checkpoint we need to defer loading the state_dict
+                # see NLPModel.on_load_checkpoint
+                checkpoint["state_dict"] = {}
+
+        # legacy checkpointing for interleaved
+        else:
```

Comment:
Can we just drop support for legacy checkpoints? We don't want to have to implement and test this scenario.

Reply:
The parent class, MegatronGPTModel still has support for legacy checkpoints in its own `on_load_checkpoint` so I copied that over.

Reply:
Yeah NeMo still supports legacy checkpoints but we decided not to support them in NeMo-Aligner. Do you see any reason to make them work with SPIN? (if yes you'll need to be sure this is working, including the loading / saving of the ref. policy).

Reply:
Okay, no worries, I have removed the legacy checkpoint code